bigtable/client/detail/bulk_mutator_test.cc

```diff
@@ -0,0 +1,379 @@
+ public:
+  MOCK_METHOD0(WaitForInitialMetadata, void());
+  MOCK_METHOD0(Finish, grpc::Status());
+  MOCK_METHOD1(NextMessageSize, bool(std::uint32_t*));
+  MOCK_METHOD1(Read, bool(::google::bigtable::v2::MutateRowsResponse*));
+};
+}  // anonymous namespace
+
+/// @test Verify that MultipleRowsMutator handles easy cases.
+TEST(MultipleRowsMutatorTest, Simple) {
+  namespace btproto = ::google::bigtable::v2;
+  namespace bt = bigtable;
+  using namespace ::testing;
+
+  // In this test we create a Mutation for two rows, which succeeds in the
+  // first RPC request.  First create the mutation.
+  bt::BulkMutation mut(
+      bt::SingleRowMutation("foo", {bigtable::SetCell("fam", "col", 0, "baz")}),
+      bt::SingleRowMutation("bar",
+                            {bigtable::SetCell("fam", "col", 0, "qux")}));
+
+  // Prepare the mocks.  The mutator should issue a RPC which must return a
+  // stream of responses,  we prepare the stream first because it is easier than
+  // to create one of the fly.
+  auto reader = absl::make_unique<MockReader>();
+  EXPECT_CALL(*reader, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse* r) {
+        {
+          auto& e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        {
+          auto& e = *r->add_entries();
+          e.set_index(1);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*reader, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  // Now prepare the client for the one request.
+  btproto::MockBigtableStub stub;
+  EXPECT_CALL(stub, MutateRowsRaw(_, _))
+      .WillOnce(Invoke(
+          [&reader](grpc::ClientContext*, btproto::MutateRowsRequest const&) {
+            return reader.release();
+          }));
+
+  auto policy = bt::DefaultIdempotentMutationPolicy();
+  bt::detail::BulkMutator mutator("foo/bar/baz/table", *policy, std::move(mut));
+
+  EXPECT_TRUE(mutator.HasPendingMutations());
+  grpc::ClientContext context;
+  auto status = mutator.MakeOneRequest(stub, context);
+  EXPECT_TRUE(status.ok());
+  auto failures = mutator.ExtractFinalFailures();
+  EXPECT_TRUE(failures.empty());
+}
+
+/// @test Verify that MultipleRowsMutator retries partial failures.
+TEST(MultipleRowsMutatorTest, RetryPartialFailure) {
+  namespace btproto = ::google::bigtable::v2;
+  namespace bt = bigtable;
+  using namespace ::testing;
+
+  // In this test we create a Mutation for two rows, one of which will fail.
+  // First create the mutation.
+  bt::BulkMutation mut(
+      bt::SingleRowMutation("foo", {bigtable::SetCell("fam", "col", 0, "baz")}),
+      bt::SingleRowMutation("bar",
+                            {bigtable::SetCell("fam", "col", 0, "qux")}));
+
+  // Prepare the mocks for the request.  First create a stream response which
+  // indicates a partial failure.
+  auto r1 = absl::make_unique<MockReader>();
```

Comment:
`s/r1/reader1/g` for readability? `r1` is such a Go-style name. ;-)

Reply:
Do you feel strongly?  I find it readable enough in this context.